File: LeetCode/1319. Number of Operations to Make Network Connected/solution.py
"""
36 / 36 test cases passed.
Runtime: 112 ms
Memory Usage: 44.5 MB
"""
class Solution:
    def makeConnected(self, n: int, connections: List[List[int]]) -> int:
        if n - 1 > len(connections):
            return -1
        used = [False] * n
        edges = [[] for _ in range(n)]
        for conn in connections:
            edges[conn[0]].append(conn[1])
            edges[conn[1]].append(conn[0])
        
        def dfs(i):
            used[i] = True
            for v in edges[i]:
                if not used[v]:
                    dfs(v)

        ans = 0
        for i in range(n):
            if not used[i]:
                dfs(i)
                ans += 1
        return ans - 1
       

File: LeetCode/1319. Number of Operations to Make Network Connected/solution.cpp
/**
 * 36 / 36 test cases passed.
 * Runtime: 128 ms
 * Memory Usage: 50.4 MB 
 */
class Solution {
public:
    void dfs(vector<vector<int>> &edges, vector<bool> &used, int u) {
        used[u] = true;
        for (int v : edges[u]) {
            if (!used[v]) {
                dfs(edges, used, v);
            }
        }
    }
    int makeConnected(int n, vector<vector<int>>& connections) {
        if (n - 1 > connections.size()) return -1;
       vector<bool> used(n);
       vector<vector<int>> edges(n);
       for (auto &conn : connections) {
           edges[conn[0]].push_back(conn[1]);
           edges[conn[1]].push_back(conn[0]);
        }

        int ans = 0;
        for (int i = 0; i < n; ++ i) {
            if (!used[i]) {
                dfs(edges, used, i);
                ++ ans;
            } 
        }
        return ans - 1;
    }
};
